pkg/buildpack/buildpack.go

```diff
@@ -336,3 +491,83 @@ func Set(exclude []string) map[string]struct{} {
 	}
 	return excludedModules
 }
+
+type ModuleTar interface {
+	Info() dist.ModuleInfo
+	Path() string
+}
+
+type moduleTar struct {
+	info   dist.ModuleInfo
+	path   string
+	writer archive.TarWriter
+}
+
+func (t *moduleTar) Info() dist.ModuleInfo {
+	return t.info
+}
+
+func (t *moduleTar) Path() string {
+	return t.path
+}
+
+func newModuleTar(dest, id, version string) (moduleTar, error) {
+	layerTar := filepath.Join(dest, fmt.Sprintf("%s.%s.tar", id, version))
+	fh, err := os.Create(layerTar)
+	if err != nil {
+		return moduleTar{}, errors.Wrapf(err, "creating file at path %s", layerTar)
+	}
+	return moduleTar{
+		info: dist.ModuleInfo{
+			ID:      id,
+			Version: version,
+		},
+		path:   layerTar,
+		writer: tar.NewWriter(fh),
+	}, nil
+}
+
+type moduleTarCollection struct {
+	rootPath string
+	modules  map[string]moduleTar
+}
+
+func newModuleTarCollection(rootPath string) *moduleTarCollection {
+	return &moduleTarCollection{
+		rootPath: rootPath,
+		modules:  map[string]moduleTar{},
+	}
+}
+
+func (m *moduleTarCollection) get(id, version string) (moduleTar, error) {
+	key := fmt.Sprintf("%s@%s", id, version)
+	if _, ok := m.modules[key]; !ok {
+		module, err := newModuleTar(m.rootPath, id, version)
+		if err != nil {
+			return moduleTar{}, err
+		}
+		m.modules[key] = module
+	}
+	return m.modules[key], nil
+}
+
+func (m *moduleTarCollection) getModules() []ModuleTar {
+	var modulesTar []ModuleTar
+	for _, v := range m.modules {
+		v := v
```

Comment:
Alternatively if you wanted to avoid this inefficiency you could change the `moduleTar` methods to be declared like `func (t moduleTar) Info() dist.ModuleInfo` (removing the pointer receiver)

Reply:
Or you could make the map `map[string]*moduleTar` ...that might be the most straightforward. But honestly I don't think it matters too much.